node/engine/engine.go

```diff
@@ -189,6 +189,8 @@ func (e *Engine) run(ctx context.Context) {
 		case pr := <-e.PaymentRequestsFromAPI:
 			res, err = e.handlePaymentRequest(pr)
 		case chainEvent := <-e.fromChain:
+			err = e.store.SetLastBlockProcessed(chainEvent.BlockNum())
```

Comment:
In this branch this is the only time we call `store.SetLastBlockProcessed`. We should improve this so that we are updating the `lastBlockProcessed` even when no events are emitted by the NitroAdjudicator contractor. This optimization would reduce the amount of processing the chainservice has to do when it is initialized and checks for any events that were emitted since the node was last online.

Reply:
Shouldn't the chainservice have a pretty smart way of checking historical events, though? If we are filtering our queries using the NitroAdjudicator address https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getlogs , we could potentially get away with only changing `lastBlockProcessed` when the adjudicator emits an event.

Reply:
Yes I think we could get away with only changing `lastBlockSeen` when the adjudicator emits an event. However, I was trying to think of a way to reduce processing time during this scenario:

** nitro node is running ***
1. Block 100 mined on-chain
2. node detects adjudicator event (AE1) in block 100
3. Block 101 mined on-chain
4. Block 102 mined on-chain
5. node processes AE1
6. node updates `lastBlockSeen` to 100
7. Block 103-20,000 mined on-chain, but no adjudicator events are emitted
8. node is turns off
9. Block 20,001-30,000 mined on-chain
10. node is restarted and begins init routines
11. node looks through blocks from `lastBlockSeen` (100) to current block (30,000) for adjudicator events

The query in step [11] to search through many blocks seems wasteful since our nitro node was actually running up until block 20,000 was mined, but `lastBlockSeen` was not updated because no adjudicator events were included in any of those blocks. Instead of searching through the block range 100-20,000, we should just be able to search through the range 20,000-30,000 since those were the only blocks mined while the node was offline.

Reply:
That's a good point, I didn't think about that scenario!

One thing to note is that ethereum makes use of [bloom filters](https://en.wikipedia.org/wiki/Bloom_filter) for logs(aka events) on [the block header.](https://www.educative.io/answers/what-is-ethereums-block-structure)  This means it's very efficient to check if a block contains an event by checking the `logsBloom` on the block header. Instead of having to parse through the a block you can just check your event topics against the bloom filter to know if the events are in the block. This means it's relatively efficient to query for events over a large range of blocks.  The [JSON-RPC API allows you to specify a from and to block](https://docs.infura.io/networks/ethereum/json-rpc-methods/eth_getlogs) so it's pretty easy to query a large range of blocks for event's we'd be interested in.

Based on that I think we can avoid worrying about this scenario too much for now. If it's something we start noticing we can revisit and implement a solution.

Reply:
That's good info on the bloom filters. I will not worry about this scenario for now but I made this issue to remind us to revisit if we see any related problems in the future: https://github.com/statechannels/go-nitro/issues/1600